### ISSUE TITLE
perf(kda_layer_norm_gated): warp-per-row float4 + 32x more blocks to fix the 32-block occupancy bound (1.4% of HBM)

### ISSUE THREAD

**Issue by xzyaoi** [opentela-ai/vkernels#144]
## Summary

`kda_layer_norm_gated` is the last kda-family kernel with an unaddressed, mechanically-diagnosable gap (issue #83 closed the D×D-state and chunked-WY items). It is **occupancy-bound**: at the recorded shape (N=8192, D=128) it launches only **32 blocks on 228 CUs** and achieves **72 GB/s = 1.4% of the 5300 GB/s HBM roof**. `docs/kernels/kda.md` already flags this as "a clear occupancy target (raise the block count or vectorise the row reduction)" — this issue tracks implementing it.

## Current measured perf + roof + gap

From `docs/kernels/kda.md` (supporting-kernels table, master table row 16, `meta/benchmarks/bench_kda.hip` / `bench_kda.sh`):

| kernel | shape | µs (med) | GB/s | binding |
|---|---|---:|---:|---|
| `kda_layer_norm_gated` | N=8192, D=128 | **174** | **72** | occupancy (32 blocks / 228 CUs) |

Byte model: `x` + `gate` reads + `out` write = `3·N·D·4 B` = 12.6 MB. At the 5300 GB/s HBM roof the kernel floor is ~2.4 µs; even a conservative 20–30% of roof puts it at ~10 µs. **Gap: ~1.4% → potential ~10–30×.**

## Root cause

`src/c/vkernels/kernels/kda.hip:51-79`: the kernel is **one thread per token row**:

```cuda
const int block = 256;
const int grid  = (N + block - 1) / block;      // N=8192 -> 32 blocks
```

- `grid = N/256 = 32` blocks on 228 CUs → ~14% of CUs get any work, ~1 wavefront total.
- Each thread then does **two serial scalar D-loops** (sum-of-squares pass, then the gated normalize/silu write) — 128 dependent scalar loads per pass, minimal memory-level parallelism per thread.

So the kernel is latency-bound on 32 blocks' serial scalar chains, not anywhere near any bandwidth limit.

## Proposed optimization

**Warp-per-row + `float4` vectorization** (D=128 → 32 `float4`s per row, exactly one 16 B load per lane):

1. One **warp per row** instead of one thread per row: a 256-thread block handles 8 rows → `grid = N/8 = 1024` blocks (≥4 full waves on 228 CUs), **32× more blocks**.
2. Load `x` and `gate` as `float4` (lane `l` takes elements `4l..4l+3` of the row); sum-of-squares via a 5-step `__shfl_xor` butterfly (warp-local — no `__syncthreads`, only `__syncwarp`).
3. Broadcast `inv` via `__shfl_sync`, write `out` as `float4` with the existing `__expf` silu.
4. `weight[d]` is shared by all rows: stage the 32 `float4`s in LDS once per block, or rely on L1/L2 broadcast — either is fine at this traffic mix.

Alternative (if a smaller diff is preferred): keep one-thread-per-row but switch both loops to `float4` (4× fewer load instructions); that alone fixes per-thread MLP but NOT the 32-block device occupancy, so the warp-per-row layout is the load-bearing part of the fix.

## Expected gain and how to measure

- Expected: 174 µs → **~10–30 µs** (6–15×), i.e. 72 GB/s → several hundred GB/s. The shape is tiny in absolute terms, but this kernel sits in the per-layer KDA path, so it compounds per layer per step.
- Measure with the existing harness: `meta/benchmarks/bench_kda.sh` already times `layer_norm_gated` standalone (bench_kda.hip:218-223) — re-run on beverin MI300A, then update the supporting-kernels table in `docs/kernels/kda.md` and row 16 of `docs/kernels-reference.md`.
- Correctness gate: `meta/benchmarks/test_kda_correct.hip` (§2, `kda_layer_norm_gated` vs CPU at K3-ish N,D).

## Risk / precedent

- **fp round-off**: the butterfly reduction reorders the sum-of-squares vs the serial CPU oracle — the same reorder concern as the (rejected for other reasons) parallel-dot experiment, which nonetheless passed `test_kda_correct` 11/11 with max_rel 1–2e-6 ≪ the 2e-2 threshold. Expect no issue; the contractive RMS denominator keeps reorder cost negligible.
- **Precedent**: this is the same fix pattern already proven elsewhere in the repo — `mhc_pre_gemm_sqrsum` (1 block → split columns, documented redesign) and `dsa_topk_logits` `split_kv` (one-wavefront occupancy → 4.7×). No rejected variant applies: the docs' kda negative results (3-barrier race, parallel dots, chunked-at-decode) all concern `kda_delta_rule_fwd`/chunked, not this kernel.

## Related (out of scope here)

The chunked-WY prefill kernel's gram phase (269 µs of 748 at H=1 S=512 D=128, ~25× above its SFU/exp floor) is the other known headroom; it is already documented as "remaining headroom (not taken)" with a named approach in `docs/kda-lds-optimization.md` and carries real numerical risk (explicit `Ainv` stability) — deliberately not bundled into this low-risk change.

**Comment by xzyaoi** [comment]
## Implemented — warp-per-row `float4` layout ✅

**Commit:** `3c76e4a` on branch `fix/issue-144` (worktree `vkernels-issue-worktrees/issue-144`)

### What changed
- `src/c/vkernels/kernels/kda.hip`: new `kda_layer_norm_gated_kernel_row` — **one warp per token row** (256-thread block → 8 rows → `N/8 = 1024` blocks, exactly as proposed). Lane `l` holds row elements `4l..4l+3` as `float4` (D=128 → one 16 B load per lane per pass); sum-of-squares via a 5-step `__shfl_xor` butterfly (full-warp, all lanes active; note this toolchain/ROCm 6.3 exposes AMD-detail `__shfl_xor`, not `__shfl_xor_sync` — same idiom as the rest of the codebase); normalize+SiLU write is also `float4`.
- The one-thread-per-row kernel is kept and auto-selected as fallback for `D % 4 != 0`; launcher picks the layout in `kda_layer_norm_gated`.
- Docs updated: `docs/kernels/kda.md` supporting-kernels section + row 16 of `docs/kernels-reference.md`.

### Correctness
- `test_kda_correct` **11/11 PASS** on MI300A (gfx942, ROCm 6.3): `layer_norm_gated` vs CPU oracle max_rel **1e-6** at N∈{8,64,8192}, D∈{128,512} — all delta_rule/cumsum/pack tests untouched and passing.
- Old-vs-new device outputs identical: max_abs_diff **3e-7**.

### Measured (beverin MI300A, nid003020, N=8192 D=128)
| variant | us(med) | GB/s | % of 5300 GB/s HBM |
|---|--:|--:|--:|
| old (thread-per-row, 32 blocks) | 170.7 | 74 | 1.4% |
| **new (warp-per-row float4, 1024 blocks)** | **4.8** | **~2600** | **~49%** |

**~35× speedup** (target was <30 µs). Method: batched event-pair timing over 1000 launches (3 reps: 4.80/4.83/4.95 µs) after the standalone per-launch event timer returned the known all-zeros broken row; cross-checked with `rocprof --stats` per-launch average 6.15 µs (job logs 641293/641303 on beverin).

### Follow-ups (not done here)
- Remaining ~2× gap to the HBM floor is weight-broadcast + latency (each warp re-reads the 32 weight float4s; an LDS-staged weight could be tried if this kernel ever shows up in profiles again).
- The kernel now sits near the event-timer floor — the repo's supporting-kernel bench may want a batched-timing mode like the A/B script used here.

**Comment by xzyaoi** [comment]
Closing per the issue's own success criteria — exceeded.

- **Result:** 174 → **4.8 µs (35×)** on MI300A (target was 10–30 µs), ~49% of HBM roof.
- Correctness: 11/11 vs CPU oracle (old/new device outputs identical to 3e-7); ROCm 6.3 all-zeros event-timer quirk cross-checked with batched 1000-launch timing + `rocprof --stats`.
- Implementation: warp-per-row float4 + 32× more blocks, per the issue.
- Commit `3c76e4a` on `fix/issue-144` (local, results comment above has full numbers); docs updated.

Follow-up push/PR/merge is intentionally not done by the agent contract.
